### PR TITLE
[ROCm] enable scaled_gemm

### PR DESCRIPTION
**Pull request by jeffdaily**
scaled_gemm for ROCm using hipblaslt.  As of ROCm 6.0, HIPBLASLT_MATMUL_DESC_AMAX_D_POINTER is not supported.  A work-around is provided, performing the absmax operation on the output buffer, but this results in some loss of accuracy for the absmax result.  For this reason the feature should be considered beta/preview.

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Comment by jianyuh** [comment]
Rebase to the trunk?

**Comment by jeffdaily** [comment]
> Rebase to the trunk?

This PR depends on https://github.com/pytorch/pytorch/pull/115214 but it got reverted due to breaking the internal build.  It's in the process of relanding.

**Comment by jeffdaily** [comment]
Manual rebase after #115214 relanded.

**Comment by jianyuh** [comment]
Hi @jeffdaily , could you rebase to the trunk? There is a conflict with master so we couldn't import it internally and test the PR. Thanks!

**Comment by jeffdaily** [comment]
> Hi @jeffdaily , could you rebase to the trunk? There is a conflict with master so we couldn't import it internally and test the PR. Thanks!

Done.

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 1]
```diff
@@ -966,10 +978,53 @@ void scaled_gemm(
...
+                all_algos[i].algo,
+                ret_workspace_size);
+        if (is_valid_status == HIPBLAS_STATUS_SUCCESS) {
+            if (ret_workspace_size <= workspaceSize) {
+                heuristicResult = all_algos[i];
+                found = true;
+                break;
+            }
+        }
+    }
```
Comment (jianyuh) [inline aten/src/ATen/cuda/CUDABlas.cpp:1024]: Ideally the above code can be encapsulated into the same `cublasLtMatmulAlgoGetHeuristic` API, so we will not diverge here.

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 2]
```diff
@@ -887,6 +889,13 @@ _scaled_mm_out_cuda(const Tensor& mat1, const Tensor& mat2,
   TORCH_CHECK(false, "_scaled_mm_out_cuda is not compiled for this platform.");
 #endif
 
+#if defined(USE_ROCM) && ROCM_VERSION >= 60000
+  // rocm's hipblaslt does not yet support amax, so calculate separately
```
Comment (jianyuh) [inline aten/src/ATen/native/cuda/Blas.cpp:893]: Are we expecting that hipblaslt will support amax in the near future (although we might use it when the output dtype is bf16/fp16 instead of fp8)?

**Review by jianyuh — COMMENTED** [review]
Is this PR ready for public / review ?

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 1 above)
Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:1024]: Currently hipblasLtMatmulAlgoGetHeuristic returns 0 solutions.  That's why I added this fallback.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 2 above)
Comment (jeffdaily) [reply aten/src/ATen/native/cuda/Blas.cpp:893]: We would need to confirm with PMs, but amax might be planned for ROCm 6.2.  This might not be soon enough for your needs.

**Review by jianyuh — APPROVED** [review]

**Comment by jeffdaily** [comment]
> @jeffdaily could you help rebase ?

Done.

**Comment by xw285cornell** [comment]
@jeffdaily it's interesting that we support Float8_e4m3fnuz on rocm, but on cuda it's Float8_e4m3fn? Just curious why this choice? Also I just run float8 benchmark in pytorch and it seems float8 is quite bad for certain cases: 

        name                shape           dtype  ref_time_s  fp8_time_s  fp8_speedup
0  attn.wqkv  (16384, 8192, 1280)  torch.bfloat16    0.000661    0.003142     0.210278
1    attn.w0  (16384, 1024, 8192)  torch.bfloat16    0.000690    0.003247     0.212516
2    ffn.w13  (16384, 8192, 7168)  torch.bfloat16    0.003471    0.016683     0.208059
3     ffn.w2  (16384, 3584, 8192)  torch.bfloat16    0.001942    0.008761     0.221673
4  attn.wqkv  (16384, 8192, 1280)   torch.float16    0.002588    0.001000     2.586915
5    attn.w0  (16384, 1024, 8192)   torch.float16    0.002339    0.001361     1.718257
6    ffn.w13  (16384, 8192, 7168)   torch.float16    0.026865    0.004075     6.593240
7     ffn.w2  (16384, 3584, 8192)   torch.float16    0.008174    0.001984     4.119691

Just a reference for H100: 
        name                shape           dtype  ref_time_s  fp8_time_s  fp8_speedup
0  attn.wqkv  (16384, 8192, 1280)  torch.bfloat16    0.000550    0.000240     2.296304
1    attn.w0  (16384, 1024, 8192)  torch.bfloat16    0.000440    0.000248     1.776945
2    ffn.w13  (16384, 8192, 7168)  torch.bfloat16    0.002994    0.001433     2.089865
3     ffn.w2  (16384, 3584, 8192)  torch.bfloat16    0.001505    0.000689     2.185172
4  attn.wqkv  (16384, 8192, 1280)   torch.float16    0.000550    0.000244     2.256594
5    attn.w0  (16384, 1024, 8192)   torch.float16    0.000458    0.000248     1.844120
6    ffn.w13  (16384, 8192, 7168)   torch.float16    0.003073    0.001430     2.149140
7     ffn.w2  (16384, 3584, 8192)   torch.float16    0.001750    0.000769     2.277292

File: test/test_matmul_cuda.py [hunk 3]
```diff
@@ -304,19 +316,19 @@ def test_float32_output_errors_with_bias(self, device) -> None:
...
         )
 
-    @unittest.skipIf(not scaled_mm_supported_device(), "FP8 is only supported on H100+ and sm_89 devices")
+    @unittest.skipIf(not scaled_mm_supported_device(), f8_msg)
     def test_float8_scale_fast_accum(self, device) -> None:
         size = (16, 16)
-        x = torch.full(size, .5, device=device, dtype=torch.float8_e4m3fn)
-        y = torch.full(size, .5, device=device, dtype=torch.float8_e5m2).t()
+        x = torch.full(size, .5, device=device, dtype=e4m3_type)
+        y = torch.full(size, .5, device=device, dtype=e5m2_type).t()
```
Comment (jianyuh) [inline test/test_matmul_cuda.py:331]: One quick check, when x and y are different dtype (e4m3 and e5m2 like this), it reports the following error:

```
File "/mnt/xarfuse/uid-169890/bbc37ff9-seed-nspid4026531836_cgpid28107062-ns-4026531841/torch_scaled_mm.py", line 9, in <module>
    out_fp8, amax_fp8 = torch._scaled_mm(x, y, use_fast_accum=True)
RuntimeError: CUDA error: HIPBLAS_STATUS_NOT_SUPPORTED when calling `HIPBLAS_STATUS_NOT_SUPPORTED
```

Is this related to the specific hipBLASLt version?

**Comment by jianyuh** [comment]
Passed the internal tests when x / y are both torch.float8_e4m3fnuz.

**Comment by jeffdaily** [comment]
> @jeffdaily it's interesting that we support Float8_e4m3fnuz on rocm, but on cuda it's Float8_e4m3fn? Just curious why this choice? Also I just run float8 benchmark in pytorch and it seems float8 is quite bad for certain cases:

The fnuz types are what our MI300 supports.  Can you share your float8 benchmark?  That would be helpful for us to tune further.

File: test/test_matmul_cuda.py (hunk 3 above)
Comment (jianyuh) [reply test/test_matmul_cuda.py:331]: In this unit test, seems that we are still checking the case with x and y in different dtype (e4m3 and e5m2) ? We should disable it if not supported?

**Comment by xw285cornell** [comment]
@jeffdaily https://github.com/pytorch-labs/float8_experimental

**Review by xw285cornell — APPROVED** [review]
Internal all good, we should be good to go!

File: test/test_matmul_cuda.py (hunk 3 above)
Comment (xw285cornell) [reply test/test_matmul_cuda.py:331]: actually I tried and both e5m2 is not supported? Agreed we should change the unit test, maybe next PR if you think this PR is already too long

**Comment by xw285cornell** [comment]
@jeffdaily maybe fix the unit test given the merge has failed :)

File: test/test_matmul_cuda.py (hunk 3 above)
Comment (jeffdaily) [reply test/test_matmul_cuda.py]: Latest commit updates the unit tests to pass with MI300 given the caveats of currently supported types.

**Comment by jeffdaily** [comment]
> @jeffdaily maybe fix the unit test given the merge has failed :)

Pushed commit to resolve the unit tests.  You'll need to redo your import to resolve the Meta Internal-Only Changes Check that is now failing due to my new commit push.

**Comment by zou3519** [comment]
@jeffdaily is there a recommended way to detect that the rocm gpu is "ROCm MI300+"?

**Comment by jeffdaily** [comment]
> @jeffdaily is there a recommended way to detect that the rocm gpu is "ROCm MI300+"?

In torch C++, we have `if (at::detail::getCUDAHooks().isGPUArch({"gfx942"}))`.

In torch python, we have `if "gfx94" in torch.cuda.get_device_properties(0).gcnArchName.split(":")[0]`.
